lib/linked_list.rb

```diff
@@ -7,39 +7,67 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if @head.nil?
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    return 0 if @head.nil?
+    count = 0
+    cur = @head
+    until cur.nil?
+      count += 1
+      cur = cur.next
+    end
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    if @head.nil?
+      @head = Node.new(data, nil)
+    else
+      cur = @head
+      while cur.next != nil
+        cur = cur.next
+      end
+      cur.next = Node.new(data, nil)
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity O(1)
   def get_last
-
+    cur = @head
```

Comment:
What if `@head` is `nil`?

```suggestion
    return nil if @head.nil?
    cur = @head
```